### PR TITLE
2619 simplify resultset filtering

### PR DESCRIPTION
**Pull request by 0biWanKenobi**
## Target

tl;dr Implementing a quicker way to filter results with a dedicated icon.

Currently, filtering a resultset by column value is a bit of a lenghty process. You must 

- click on a cell ( or the IN.. filter won't work)
- left click to open context menu
- choose submenu filters
- select IN..
- proceed with filtering

As proposed in issue #2619, I implemented a pop up menu with the list of a column distinct values. The user is able to open it via a new icon, place side by side with the sorting icon. This means that the user can filter like this:

- click filter icon
- choose cell value

and that's it! Much quicker ;)

image

## Implementation

I added a new icon alongside the sorting one: click detection is handled exactly the same way. Event triggering, too, happens the same. 

image

The list of column values is a `Composite` holding a `TableViewer` with gray bg and no lines or headers, so that it looks like a menu. Under the list there is a textbox to filter results. 
The list of values is obtained the same way as with `FilterValueEditDialog` class, only in this case the `TableViewer` has just the labels column; this way it's more lightweight for the user. 

## Major project changes

Only one really: I refactored `FilterValueEditDialog` in a new package called `org.jkiss.dbeaver.ui.controls.resultset.valuefilter`. This has been done because there's a lot of common code between the new menu and the existing dialog used for the "IN.." filter, so I wanted to avoid duplicate code and prevent methods and variables from being exposed to all classes in the previous package. Aside from moving code from place to place, no breaking changes have been done and the preexisting code works exactly the same as before. 

## Implementation details

Modified resources:

`GridColumn`, `GridColumnRenderer`, `LightGrid`, `Spreadsheet`, `SpreadsheetPresentation`, `IResultSetController`, `ResultSetViewer`, `FilterValueEditMenu`, `FilterValueEditDialog`, `GenericFilterValueEdit`.

- _GridColumn_: 
  - Since there's a new icon, increase header height to avoid cluttering;
  - With the same logic of the sort icon mouse position detection, add detection for the new filter icon;
  - Adjust sort arrow detection for new x coordinate.

- _GridColumnRenderer_:
   - Since there's a new icon, increase header height to avoid cluttering;
  - Define icon image and size;
  - in paint(), adjust header width for text considering new icon; compute bounds a draw new icon.

- _LightGrid_:
  - With the same logic of the sort icon click detection, add detection for the new filter icon;
  - When a click happens, trigger new Event "Event_FilterColumn".

- _Spreadsheet_:
  - Let Eclipse reorganize imports;
  - Added listener for new event triggered from LightGrid by filter icon click;
  - Call ResultSetViewer's method thru SpreadSheetPresentation when click event is handled.

- _SpreadsheetPresentation_:
  - Before calling ResultSetViewer, deselects any cell in current or other columns 

- _IResultSetController:
  - Abstract method to call when user clicks filter icon.

- _ResultSetViewer_:
  - Concrete implementation of interface method to handle user click;
  - Uses new class FilterValueEditMenu to show a list of column values.

- _FilterValueEditMenu_:
  - New class.  Since it had a lot in common with FilterValueEditDialog, a new package has been created, along with a generic class holding common code;
Package name is `org.jkiss.dbeaver.ui.controls.resultset.valuefilter`;
  - Uses TableView without checkboxes, it allows single selection only;
  - Is modal, user can use Esc button to close it without filtering the resultset.

- _FilterValueEditDialog_:
  - Has been moved to `org.jkiss.dbeaver.ui.controls.resultset.valuefilter`;
  - Common methods have been moved to a generic class;
  - Works exactly the same as before.

- _GenericFilterValueEdit_:
  - Holds all common code and variables between `FilterValueEditMenu` and `FilterValueEditDialog`;

### CODE REVIEW

**Comment by serge-rider** [comment]
Wow, great job.
I will review this shortly (hopefully on Monday)

**Comment by serge-rider** [comment]
Found a couple of issues:

1. Filter drop-down fails with NPE when you click on the filter icon second time (click, select value, click on the same column icon again):
```java
java.lang.NullPointerException
	at org.jkiss.dbeaver.ui.controls.resultset.valuefilter.FilterValueEditMenu.createMultiValueSelector(FilterValueEditMenu.java:135)
	at org.jkiss.dbeaver.ui.controls.resultset.valuefilter.FilterValueEditMenu.init(FilterValueEditMenu.java:59)
	at org.jkiss.dbeaver.ui.controls.resultset.valuefilter.FilterValueEditMenu.<init>(FilterValueEditMenu.java:41)
	at org.jkiss.dbeaver.ui.controls.resultset.ResultSetViewer.showDistinctFilter(ResultSetViewer.java:1598)
	at org.jkiss.dbeaver.ui.controls.resultset.spreadsheet.SpreadsheetPresentation.showFiltering(SpreadsheetPresentation.java:1062)
	at org.jkiss.dbeaver.ui.controls.resultset.spreadsheet.Spreadsheet.handleEvent(Spreadsheet.java:324)
```
2. It is better to make drop-down like composite (FilterValueEditMenu) as non-modal dialog. Similarly to how schema selector dialog was made (it sets focus in selector table and it closes when table looses focus): see DataSourceManagementToolbar.showDatabaseSelector and SelectObjectDialog.java . It is much more user friendly.


Also it makes sense to add standard copyright header in all new source files.

Or I can do this after PR merge. And add some other extra improvements as: shortcut, filter icon width reduce, last filter value selection save and maybe some others.

**Comment by 0biWanKenobi** [comment]
I'll take a look at the NPE and make the drop-down close when focus is lost. Sorry about forgetting the copyright header, I'll add that too! 
I'll leave the improvements to you, if that's ok with you 😄

**Comment by 0biWanKenobi** [comment]
I can't seem to reproduce the bug you mention when clicking the same filter icon multiple times. It happened once, seemingly at random, but I wasn't debugging.. 😕  can you help with some more details?

**Comment by serge-rider** [comment]
NPE is gone 👍 
But I still think that drop-down should be a dialog. I'll change this by myself, 
thanks a lot for contribution!

**Comment by 0biWanKenobi** [comment]
You're welcome! It's been fun 😄  I was thinking, maybe we should make the dropdown list only visible cells? It would help composite filtering a lot!

I went through the code and realized that the whole `KeyLoadJob` purpose in `GenericFilterValueEdit` is to retrieve the values for all cells, even those hidden by the current filter.. Maybe it could be avoided altogether?

**Comment by serge-rider** [comment]
Yep, perhaps. Also I think it should allow multiple values selection (with checboxes).
Please check my changes (non-modal dialog and others)

**Comment by 0biWanKenobi** [comment]
Yeah I've seen them, great work! After your changes everything is much more refined. I had totally missed the integration with the filtering option in DBeaver settings 😅   

Regarding multiple selection, it would bring in the need of some sort of confirmation, and I don't really like the idea of introducing buttons. So I implemented ctrl+click for multi selection (see video in attached zip), how do you like it? Could be a good idea?

[Demo multiselection.zip](https://github.com/dbeaver/dbeaver/files/1640648/Demo.multiselection.zip)

As a personal preference, I would have the dropdown look like this, without the header..
image
Because I feel it's more of a menu than a dialog.. but then again, it's just my POV!

**Comment by esequiel-ferreira** [comment]
nice feature! Its very useful. I would suggest a little improvement: When there are many possible values, the popup shows a textbox for search in the possible values. 

Could you guys set focus to textbox when its visible? Also, when press Up key inside this textbox, it would be nice to set focus to Grid of possible values, so we would be able to navigate then with arrows keys.

The point is: it would great if I was able to do all the process with keyboard after clicking in the Filter button.

**Comment by serge-rider** [comment]
@esequiel-ferreira good point. Check it in the next version.

**Comment by esequiel-ferreira** [comment]
Nice! 

Just one more thing: When I run a select statement with alias, the filter button is still there and when clicking on it a popup with possible values is shown. However, when applying a value to filtering an error happens. 

Its clear from #1748 that filtering alias column is currently not possible, but to get this feature (filter column popup) at state of art, i think this issue should be handled.
